File: android/src/main/java/com/nami/reactlibrary/NamiPaywallManagerBridgeModule.java
package com.nami.reactlibrary;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.util.Log;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.Callback;


import com.namiml.BuildConfig;
import com.namiml.Nami;
import com.namiml.NamiConfiguration;
import com.namiml.NamiLogLevel;
import com.namiml.NamiPaywallManager;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;


public class NamiPaywallManagerBridgeModule extends ReactContextBaseJavaModule {
    private final ReactApplicationContext reactContext;

    public NamiPaywallManagerBridgeModule(ReactApplicationContext reactContext) {
        super(reactContext);
        this.reactContext = reactContext;
    }

    @NotNull
    @Override
    public String getName() {
        return "NamiPaywallManagerBridge";
    }

    @ReactMethod
    public void raisePaywall() {
//        [[NamiPaywallManager shared] raisePaywallFromVC:nil];

        final Activity activity = getCurrentActivity();
        Log.e("ReactNative", "Nami Activity to raise paywall is " + activity.toString());

        if (NamiPaywallManager.canRaisePaywall()) {
            Log.e("ReactNative", "NAMI - About to raise Paywall ");
            if (activity != null) {
                Log.e("ReactNative", "Raising Paywall: ");
                NamiPaywallManager.raisePaywall(activity, false);
            } else {
                Log.e("ReactNative", "Activity from react getCurrentActivity was null. ");
            }
        } else {
            Log.e("ReactNative", "Paywall not raised, SDK says paywall cannot be raised at this time. ");
        }
    }
    @ReactMethod
    public void canRaisePaywall(Promise p) {
//        BOOL canRaise = [[NamiPaywallManager shared] canRaisePaywall];
//        completion(@[[NSNumber numberWithBool:canRaise]]);

        ArrayList<Boolean> canRaiseResult = new ArrayList<Boolean>();

        canRaiseResult.add(NamiPaywallManager.canRaisePaywall());
        p.resolve(canRaiseResult);
    }

}

File: android/src/main/java/com/namiml/reactnative/NamiPaywallManagerBridgeModule.kt
package com.namiml.reactnative

import android.app.Activity
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.turbomodule.core.interfaces.TurboModule
import com.namiml.paywall.NamiPaywallManager
import com.namiml.paywall.NamiSKU
import com.namiml.paywall.model.NamiPurchaseSuccess

@ReactModule(name = NamiPaywallManagerBridgeModule.NAME)
class NamiPaywallManagerBridgeModule internal constructor(
    reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext), TurboModule {

    companion object {
        const val NAME = "RNNamiPaywallManager"
    }

    override fun getName(): String = NAME

    private var latestPaywallActivity: Activity? = null

    @ReactMethod
    fun buySkuComplete(dict: ReadableMap) {
        val product = dict.getMap("product")
        val productId = product?.getString("id")
        val skuRefId = product?.getString("skuId")
        val typeString = product?.getString("type")
        val storeType = dict.getString("storeType") ?: "GooglePlay"
        var purchaseSuccess: NamiPurchaseSuccess? = null

        if (productId != null && skuRefId != null) {
            val namiSku = NamiSKU.create(
                skuRefId = skuRefId,
                skuId = productId,
            )

            when (storeType) {
                "GooglePlay" -> {
                    val purchaseToken = dict.getString("purchaseToken")
                    val orderId = dict.getString("orderId")
                    if (purchaseToken != null && orderId != null) {
                        purchaseSuccess = NamiPurchaseSuccess.GooglePlay(
                            product = namiSku,
                            orderId = orderId,
                            purchaseToken = purchaseToken,
                        )
                    }
                }
                "Amazon" -> {
                    val receiptId = dict.getString("receiptId")
                    val localizedPrice = dict.getString("localizedPrice")
                    val userId = dict.getString("userId")
                    val marketplace = dict.getString("marketplace")
                    if (receiptId != null && localizedPrice != null && userId != null && marketplace != null) {
                        purchaseSuccess = NamiPurchaseSuccess.Amazon(
                            product = namiSku,
                            receiptId = receiptId,
                            localizedPrice = localizedPrice,
                            userId = userId,
                            marketplace = marketplace,
                        )
                    }
                }
            }

            if (purchaseSuccess != null) {
                val activity = latestPaywallActivity ?: reactApplicationContext.getCurrentActivity()
                if (activity != null) {
                    NamiPaywallManager.buySkuComplete(activity, purchaseSuccess)
                } else {
                    Log.w(NAME, "No activity available to complete purchase.")
                }
            } else {
                Log.d(NAME, "Unable to create a valid NamiPurchaseSuccess object. Purchase not completed.")
            }
        }
    }

    @ReactMethod
    fun registerCloseHandler() {
        NamiPaywallManager.registerCloseHandler { activity ->
            latestPaywallActivity = activity
            val map = Arguments.createMap().apply {
                putBoolean("paywallCloseRequested", true)
            }
            emitEvent("PaywallCloseRequested", map)
        }
    }

    @ReactMethod
    fun dismiss(promise: Promise) {
        latestPaywallActivity?.let {
            NamiPaywallManager.dismiss(it) { result -> promise.resolve(result) }
        } ?: promise.resolve(false)
    }

    @ReactMethod
    fun registerBuySkuHandler() {
        NamiPaywallManager.registerBuySkuHandler { activity, sku ->
            latestPaywallActivity = activity
            try {
                val dictionary = sku.toSkuDict()
                emitEvent("RegisterBuySKU", dictionary)
            } catch (e: Exception) {
                Log.e(NAME, "Failed to convert SKU: ${e.localizedMessage}")
            }
        }
    }

    @ReactMethod
    fun registerSignInHandler() {
        NamiPaywallManager.registerSignInHandler { activity ->
            latestPaywallActivity = activity
            val map = Arguments.createMap().apply {
                putBoolean("paywallSignInRequested", true)
            }
            emitEvent("PaywallSignInRequested", map)
        }
    }

    @ReactMethod
    fun registerRestoreHandler() {
        NamiPaywallManager.registerRestoreHandler { activity ->
            latestPaywallActivity = activity
            val map = Arguments.createMap().apply {
                putBoolean("paywallRestoreRequested", true)
            }
            emitEvent("PaywallRestoreRequested", map)
        }
    }

    @ReactMethod
    fun registerDeeplinkActionHandler() {
        NamiPaywallManager.registerDeepLinkHandler { activity, url ->
            latestPaywallActivity = activity
            emitEvent("PaywallDeeplinkAction", url)
        }
    }

    @ReactMethod
    fun show() {
        // No-op on Android
    }

    @ReactMethod
    fun hide() {
        // No-op on Android
    }

    @ReactMethod
    fun isHidden(promise: Promise) {
        promise.resolve(false)
    }

    @ReactMethod
    fun isPaywallOpen(promise: Promise) {
        promise.resolve(NamiPaywallManager.isPaywallOpen())
    }

    @ReactMethod
    fun buySkuCancel() {
        reactApplicationContext.runOnUiQueueThread {
            latestPaywallActivity?.let {
                NamiPaywallManager.buySkuCancel(it)
            } ?: reactApplicationContext.getCurrentActivity()?.let {
                NamiPaywallManager.buySkuCancel(it)
            } ?: NamiPaywallManager.buySkuCancel()
        }
    }

    @ReactMethod
    fun setProductDetails(productDetails: String, allowOffers: Boolean) {
        NamiPaywallManager.setProductDetails(productDetails, allowOffers)
    }

    @ReactMethod
    fun setAppSuppliedVideoDetails(url: String, name: String?) {
        NamiPaywallManager.setAppSuppliedVideoDetails(url, name)
    }

    @ReactMethod
    fun allowUserInteraction(allowed: Boolean) {
        NamiPaywallManager.allowPaywallInteraction(allow = allowed)
    }

    @ReactMethod fun addListener(eventName: String?) {}
    @ReactMethod fun removeListeners(count: Int?) {}

    private fun emitEvent(name: String, payload: Any?) {
        val emitter = reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
        emitter.emit(name, payload)
    }
}
